string id
float64 grad
float64 hessian
bool done
